### ISSUE TITLE
[Feature request] Summarized information on PR Comments

### ISSUE THREAD

**Issue by Zeitsperre** [green-coding-solutions/eco-ci-energy-estimation#107]
I'm wondering if there's interest in providing a more aggregated summary of information when commenting the results on a Pull Request.

Currently, if one sets `task: display-results; pr-comment: true`, the particular *task* will generate a comment, but in the case of a more complex/multi-stage CI pipeline, it would be more illuminating to know the full information, based more on the summary of all steps/tasks.

In my current CI pipeline, the idea would resemble

```
Step0[code linting] ⇾ Step1[testing](multiple VMs in parallel) ⇾ Step2[finish](PR comment == True)
```

The information I'd like to see posted in a pull request would be the totals for the last run, updated with each commit (perhaps using the https://github.com/marketplace/actions/find-comment Action). This info is currently visible on the metrics.green-coding.io dashboard.

**Comment by ArneTR** [comment]
I agree, it would be nice to have this in the PR also. Our intention though is to not duplicate functionality and it is way easier displaying this in a proper web interface on a separate server as we currently do it.

At the moment what you are truly looking for (at least I believe :) ) to compare a commit with the one before is not really doable on GitHub Actions anyway.
We posted a longer piece of the inherent variability of GitHub Actions due to the shared environment: https://www.green-coding.io/case-studies/ci-pipeline-energy-variability/

So in order to understand if your commit really changed something in terms of energy it's effect must be bigger than the StdDev of the pipeline, which is sadly quite high on GitHub Actions.

The way to mitigate this would be to use a custom runner. We plan to release that in the coming months. So if that is interesting for you stay tuned :)

In case I missed something or misunderstood your question please follow up

**Comment by Zeitsperre** [comment]
@ArneTR 

Thanks for getting back to me.

That's not exactly what I had in mind, but I definitely did notice that the inter-run variability for energy use and carbon is quite high and uncertain. I agree that there wouldn't be much value comparing carbon estimates between runs (how could one even specify to GitHub that they should use less carbon intensive runners? That would be an interesting feature though. Maybe something to consider asking them directly).

I think what I would like to be able to have posted on a Pull Request or in the Workflow summary is a running total of energy and estimated emissions for that Pull Request, i.e. the information coming from the `totals` of all runs/steps:
![image](https://github.com/user-attachments/assets/6c5e7694-3c7f-4873-be27-f093a2d7e574)

I would be curious to know what the average estimated impact of the CI runs for a Pull Request (and perhaps the CI running on pushed to `main`) have contributed overall to the carbon emissions associated with a project/organization. If this information could be presented in a badge, there would be added pressure to optimize existing CI pipelines and perhaps dissuade developers from triggering the full CI pipeline as often.

I'm not sure how long the information is retained, but having lifetime tallies can help a lot with tracking this information, and aid in justifying energy saving and carbon mitigation/sequestration efforts (or carbon offsets, [effective or not](https://www.cbc.ca/radio/whatonearth/carbon-offsets-might-be-a-dangerous-distraction-from-more-effective-climate-action-experts-say-1.5946764)).

I really like the idea of this metric. Would love to see this adopted into more projects!

(Also, the reason I mentioned the `find-comment` action is because this comment showing the running totals could be replaced after every CI run (similar to other actions, like those that examine code coverage) in order to not create a bunch of comments.)

**Comment by ArneTR** [comment]
I just made a PR to discuss possible display.

How do you like this? 

https://github.com/green-coding-solutions/green-metrics-tool/pull/998

It can then also be brought into the PR view

**Comment by Zeitsperre** [comment]
That looks very useful indeed! If it's possible to keep total estimates around for a significant amount of time, having access to badges with some very basic stats would be perfect. Something like the following would be helpful for tracking:

- {Average | Total} {power consumption | emissions}/run for all runs
- {Average | Total} {power consumption | emissions}/run for the last {3 | 6 | 9 | 12} months

I imagine that comparing the averages would help illuminate whether efforts are being made (by the developer) to reduce energy usage in the CI configuration or decarbonize (this being more within the control of GitHub).

Having something that can be placed in a ReadMe and set up to track metrics for the `main` branches would be quite informative, thanks for looking into this!

**Comment by ArneTR** [comment]
Hey @Zeitsperre 

Holiday season finally allowed for some further time to work on this :)

I really like your recommendations and have implemented them in Eco-CI and also in the API endpoint on the Green Metrics Tool.

Find an example of a live display here: https://metrics.green-coding.io/ci.html?repo=green-coding-solutions/ci-carbon-testing&branch=main&workflow=48163287&start_date=2021-12-15&end_date=2024-12-22

<img width="995" alt="Screenshot 2024-12-22 at 1 09 29 PM" src="https://github.com/user-attachments/assets/c1b204a9-7815-4d39-8659-f8424646d4b0" />


The running total of the pipeline will now also shown in the PR. Great recommendation!

Find it here as an example: https://github.com/green-coding-solutions/green-metrics-tool/pull/1027
<img width="656" alt="Screenshot 2024-12-22 at 1 21 01 PM" src="https://github.com/user-attachments/assets/23d6a590-fc9b-47ae-b677-1134264cb00a" />


Furtermore I have implemented an average. Since it has no use in PR views and only really makes sense on the main branch or very long open PRs I have so far not displayed it anywhere.

However, this is the endpoint for now:
- https://api.green-coding.io/v1/ci/badge/get?repo=green-coding-solutions/ci-carbon-testing&branch=main&workflow=48163287&mode=avg&duration_days=30
- <img src="https://api.green-coding.io/v1/ci/badge/get?repo=green-coding-solutions/ci-carbon-testing&branch=main&workflow=48163287&mode=avg&duration_days=30">

I would like to use it internally for a while before coining it official and documenting it.

**Comment by Zeitsperre** [comment]
Happy New Year @ArneTR! 

This looks fantastic, exactly what I had in mind! I'll likely update one or two of my projects to display these new statistics. I'm certain my team (and hopefully others) will find this really fascinating!

Thanks so much for spending time on this. Again, this is such a good initiative!

**Comment by ArneTR** [comment]
Nice! I just created a new release and updated the `@v4` version.

Will close this issue then. Thanks for the idea!
